File: 0198-house-robber/0198-house-robber.java
class Solution {

    //RECURANCE
    private int recurse(int idx, int[] nums, int[] dp){
        if(idx<0) return 0;

        if(idx==0) return nums[0];

        if(dp[idx]!=-1) return dp[idx];

        int pick = nums[idx] + recurse(idx-2, nums, dp);
        int not = 0 + recurse(idx-1, nums, dp);
        
        return dp[idx] = Math.max(pick, not);
    }

    // private int recurse(int idx, int[] nums, int[] dp){
    //     if(idx<0) return 0;

    //     if(idx==0) return nums[0];
        
    //     if(dp[idx]!=-1) return dp[idx];
    //     int pick = nums[idx] + recurse(idx-2, nums, dp);
    //     int not = 0 + recurse(idx-1, nums, dp);

    //     return dp[idx] = Math.max(pick, not);
    // }

    public int rob(int[] nums) {
        int[] dp = new int[nums.length];
        Arrays.fill(dp, -1);
        return recurse(nums.length-1, nums, dp);
    }
}

File: 0198-house-robber/0198-house-robber.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums)==1:
            return nums[0]

        n = len(nums)+1
        dp = [-1 for _ in range(len(nums)+1)]
        dp[n-1] = 0

        for i in range(n-2, -1, -1):
            dp[i] = max(nums[i] + dp[min(i+2, n-1)], dp[i+1])

        return dp[0]